File: decompiler/cpp/kuna_assert.cc
/* ###
 * kuna-owned: the typed assertion API (see kuna_assert.hh).
 */
#include "kuna_assert.hh"
#include "grammar.hh"

namespace ghidra {

/// The session log of applied assertions (console-lifetime; survives
/// function clears the same way the Override store does).
static vector<KunaAssertion> assertLog;

/// Read the rest of the stream as whitespace-separated tokens
static void tokenizeRest(istream &s,vector<string> &tokens)

{
  s >> ws;
  while(!s.eof()) {
    string word;
    s >> word >> ws;
    if (word.empty()) break;
    tokens.push_back(word);
  }
}

/// Re-join tokens [start,end) into a single space-separated string
static string joinTokens(const vector<string> &tokens,int4 start,int4 end)

{
  string res;
  for(int4 i=start;i<end;++i) {
    if (i != start) res += ' ';
    res += tokens[i];
  }
  return res;
}

void IfcKunaAssert::applyOption(const string &optname,const string &p1,
				const string &p2,const string &p3)
{
  string res = dcp->conf->options->set(ElementId::find(optname,0),p1,p2,p3);
  *status->optr << res << endl;
}

/// \class IfcKunaAssert
/// \brief Write a typed assertion: `kassert <stage> <substage> <args...> [hard|hint]`
///
/// Validates the stage/sub-stage pair against the registry catalog, then
/// dispatches to the existing store for that sub-stage.  Records the
/// assertion (with its computed minimal rewind stage) in the session log.
void IfcKunaAssert::execute(istream &s)

{
  if (dcp->conf == (Architecture *)0)
    throw IfaceExecutionError("No load image present");
  string stagecode;
  s >> ws >> stagecode;
  if (stagecode == "list") {
    listAssertions();
    return;
  }
  KunaStage stage;
  if (!kunaStageFromCode(stagecode,stage))
    throw IfaceParseError("Bad stage code (expecting P0 or S1..S9): "+stagecode);
  string subname;
  s >> ws >> subname;
  if (subname.empty())
    throw IfaceParseError("Missing sub-stage name (see `stage list`)");
  const KunaSubStage *sub = kunaLookupSubStage(subname);
  if (sub == (const KunaSubStage *)0)
    throw IfaceParseError("Unknown sub-stage: "+subname+" (see `stage list`)");
  if (sub->stage != stage)
    throw IfaceParseError("Sub-stage "+subname+" belongs to stage "+
			  string(kunaStageCode(sub->stage))+", not "+stagecode);

  vector<string> tokens;
  tokenizeRest(s,tokens);

  KunaStrength requested = sub->strength;	// Default to the catalog strength
  if (!tokens.empty()) {			// Optional trailing hard|hint
    if (tokens.back() == "hard") { requested = kstrength_hard; tokens.pop_back(); }
    else if (tokens.back() == "hint") { requested = kstrength_hint; tokens.pop_back(); }
  }

  KunaAssertion rec;
  rec.funcName = (dcp->fd != (Funcdata *)0) ? dcp->fd->getName() : "(global)";
  rec.stage = stage;
  rec.substage = subname;
  rec.args = joinTokens(tokens,0,tokens.size());
  rec.strength = requested;
  rec.applied = sub->strength;	// Underlying Ghidra stores are uniformly HARD mechanisms
  rec.rewind = sub->rewind;

  // Dispatch table: sub-stage -> existing store.  Function-scoped stores
  // require a current function, exactly like the legacy commands they wrap.
  string sname(subname);
  istringstream rest(rec.args);
  int4 discard;
  if (sname == "edge-virtualization") {		// -> Override::insertForceGoto
    if (dcp->fd == (Funcdata *)0)
      throw IfaceExecutionError("No function selected");
    if (tokens.size() != 2)
      throw IfaceParseError("kassert S7 edge-virtualization <branch-addr> <dest-addr>");
    rest >> ws;
    Address target(parse_machaddr(rest,discard,*dcp->conf->types));
    rest >> ws;
    Address dest(parse_machaddr(rest,discard,*dcp->conf->types));
    dcp->fd->getOverride().insertForceGoto(target,dest);
  }
  else if (sname == "dead-definition-gate") {	// -> Override::insertDeadcodeDelay
    if (dcp->fd == (Funcdata *)0)
      throw IfaceExecutionError("No function selected");
    if (tokens.size() != 2)
      throw IfaceParseError("kassert S3 dead-definition-gate <space> <delay>");
    AddrSpace *spc = dcp->conf->getSpaceByName(tokens[0]);
    if (spc == (AddrSpace *)0)
      throw IfaceParseError("Bad space: "+tokens[0]);
    istringstream ds(tokens[1]);
    int4 delay = -1;
    ds >> delay;
    if (delay < 0)
      throw IfaceParseError("Need non-negative delay integer");
    dcp->fd->getOverride().insertDeadcodeDelay(spc,delay);
  }
  else if (sname == "switch-model") {		// -> Override::insertMultistageJump
    if (dcp->fd == (Funcdata *)0)
      throw IfaceExecutionError("No function selected");
    if (tokens.size() != 1)
      throw IfaceParseError("kassert S2 switch-model <branchind-addr>");
    rest >> ws;
    Address addr(parse_machaddr(rest,discard,*dcp->conf->types));
    dcp->fd->getOverride().insertMultistageJump(addr);
  }
  else if (sname == "flow-classification") {	// -> Override::insertFlowOverride
    if (dcp->fd == (Funcdata *)0)
      throw IfaceExecutionError("No function selected");
    if (tokens.size() != 2)
      throw IfaceParseError("kassert S2 flow-classification <addr> <branch|call|callreturn|branchind|return>");
    rest >> ws;
    Address addr(parse_machaddr(rest,discard,*dcp->conf->types));
    uint4 type = Override::stringToType(tokens[1]);
    if (type == Override::NONE)
      throw IfaceParseError("Bad flow override type: "+tokens[1]);
    dcp->fd->getOverride().insertFlowOverride(addr,type);
  }
  else if (sname == "prototype-source") {	// -> FuncProto input/output locks
    if (dcp->fd == (Funcdata *)0)
      throw IfaceExecutionError("No function selected");
    if (tokens.size() != 1 || (tokens[0] != "lock" && tokens[0] != "unlock"))
      throw IfaceParseError("kassert S4 prototype-source <lock|unlock>");
    bool val = (tokens[0] == "lock");
    dcp->fd->getFuncProto().setInputLock(val);
    dcp->fd->getFuncProto().setOutputLock(val);
  }
  else if (sname == "type-propagation") {	// -> retype + typelock
    if (dcp->fd == (Funcdata *)0)
      throw IfaceExecutionError("No function selected");
    if (tokens.size() < 2)
      throw IfaceParseError("kassert S5 type-propagation <symbol> <type-declaration>");
    string symname = tokens[0];
    string newname;
    istringstream ts(joinTokens(tokens,1,tokens.size())+" "+symname);
    Datatype *ct = parse_type(ts,newname,dcp->conf);
    vector<Symbol *> symList;
    dcp->readSymbol(symname,symList);
    if (symList.empty())
      throw IfaceExecutionError("No symbol named: "+symname);
    if (symList.size() > 1)
      throw IfaceExecutionError("More than one symbol named: "+symname);
    Symbol *sym = symList[0];
    if (sym->getCategory() == Symbol::function_parameter)
      dcp->fd->getFuncProto().setInputLock(true);
    sym->getScope()->retypeSymbol(sym,ct);
    sym->getScope()->setAttribute(sym,Varnode::typelock);
  }
  else if (sname == "merge-aggressiveness") {	// -> Symbol::setIsolated
    if (dcp->fd == (Funcdata *)0)
      throw IfaceExecutionError("No function selected");
    if (tokens.size() != 1)
      throw IfaceParseError("kassert S6 merge-aggressiveness <symbol>");
    vector<Symbol *> symList;
    dcp->readSymbol(tokens[0],symList);
    if (symList.empty())
      throw IfaceExecutionError("No symbol named: "+tokens[0]);
    if (symList.size() > 1)
      throw IfaceExecutionError("More than one symbol named: "+tokens[0]);
    symList[0]->setIsolated(true);
  }
  else if (sname == "naming-policy") {		// -> rename + namelock
    if (dcp->fd == (Funcdata *)0)
      throw IfaceExecutionError("No function selected");
    if (tokens.size() != 2)
      throw IfaceParseError("kassert S9 naming-policy <oldname> <newname>");
    vector<Symbol *> symList;
    dcp->readSymbol(tokens[0],symList);
    if (symList.empty())
      throw IfaceExecutionError("No symbol named: "+tokens[0]);
    if (symList.size() > 1)
      throw IfaceExecutionError("More than one symbol named: "+tokens[0]);
    Symbol *sym = symList[0];
    sym->getScope()->renameSymbol(sym,tokens[1]);
    sym->getScope()->setAttribute(sym,Varnode::namelock);
  }
  else if (sname == "comparison-canonicalization") {	// -> option compareform
    if (tokens.size() != 1)
      throw IfaceParseError("kassert S3 comparison-canonicalization <canonical|original>");
    applyOption("compareform",tokens[0],"","");
  }
  else if (sname == "pointer-notation") {	// -> option arraynotation
    if (tokens.size() != 1)
      throw IfaceParseError("kassert S9 pointer-notation <on|off>");
    applyOption("arraynotation",tokens[0],"","");
  }
  else if (sname == "simplification-quiescence") {	// -> option togglerule
    if (tokens.size() != 2)
      throw IfaceParseError("kassert S3 simplification-quiescence <rulename> <on|off>");
    applyOption("togglerule",tokens[0],tokens[1],"");
  }
  else if (sname == "readability-rewrites") {	// -> option analyzeforloops
    if (tokens.size() != 1)
      throw IfaceParseError("kassert S8 readability-rewrites <on|off>");
    applyOption("analyzeforloops",tokens[0],"","");
  }
  else if (sname == "extrapop") {		// -> option extrapop
    if (tokens.size() < 1 || tokens.size() > 2)
      throw IfaceParseError("kassert S4 extrapop <value> [<function>]");
    applyOption("extrapop",tokens[0],(tokens.size()>1) ? tokens[1] : "","");
  }
  else if (sname == "pipeline-variant") {	// -> option setaction
    if (tokens.size() < 1 || tokens.size() > 2)
      throw IfaceParseError("kassert P0 pipeline-variant <root-action> [<group>]");
    applyOption("setaction",tokens[0],(tokens.size()>1) ? tokens[1] : "","");
  }
  else {
    if (sub->latent)
      throw IfaceExecutionError("Sub-stage "+sname+" is LATENT: no override surface exists yet (kuna roadmap)");
    throw IfaceExecutionError("Sub-stage "+sname+" is not yet routable through kassert; use its native surface: "+sub->exposure);
  }

  assertLog.push_back(rec);
  *status->optr << "kassert applied: [" << kunaStageCode(rec.stage) << "] " << rec.substage;
  if (!rec.args.empty())
    *status->optr << ' ' << rec.args;
  *status->optr << "  rewind->" << kunaStageCode(rec.rewind)
		<< " (Ghidra-actual: whole-function)" << endl;
}

/// Dump the session's typed assertions: `kassert list`
void IfcKunaAssert::listAssertions(void)

{
  ostream &os( *status->fileoptr );	// bulk stream: assertable from datatests
  if (assertLog.empty()) {
    os << "No kuna assertions recorded this session" << endl;
    return;
  }
  os << "Recorded assertions (rewind scope is REPORTED per stage-model.md s12; "
     << "Ghidra-actual re-run remains whole-function):" << endl;
  for(int4 i=0;i<(int4)assertLog.size();++i) {
    const KunaAssertion &rec( assertLog[i] );
    os << "  " << dec << i << ": [" << rec.funcName << "] "
       << kunaStageCode(rec.stage) << '/' << rec.substage;
    if (!rec.args.empty())
      os << ' ' << rec.args;
    os << "  strength=" << ((rec.strength == kstrength_hard) ? "HARD" :
			    (rec.strength == kstrength_hint) ? "HINT" : "-");
    if (rec.strength == kstrength_hint && rec.applied == kstrength_hard)
      os << " (applied via HARD mechanism)";
    os << "  rewind->" << kunaStageCode(rec.rewind) << endl;
  }
}

} // End namespace ghidra

File: decompiler/crates/kuna-decomp/src/p0_knowledge/kuna_assert.rs
//! Port of `decompiler/cpp/kuna_assert.{cc,hh}` — the typed assertion API
//! (`docs/history/stage-model.md` section 12).
//!
//! `kassert` is a thin facade over the existing battle-tested stores (Override,
//! Symbol-DB locks, FuncProto locks, the option database): it validates a
//! `<stage> <substage>` pair against the stage registry ([`kuna_phases`]) and
//! dispatches to the store for that sub-stage. What it ADDS is the model —
//! every applied assertion is recorded with its stage, sub-stage, strength, and
//! the **computed minimal rewind scope** (`docs/history/stage-model.md` section 12), which
//! `kassert list` REPORTS (Ghidra-actual re-run remains whole-function).
//!
//! ## What this module owns vs. STUB(W5)
//!
//! The C++ `IfcKunaAssert::execute` is a console command bound to
//! `IfaceDecompData` (the `dcp->conf`/`dcp->fd`, the `Override`/`FuncProto`/
//! symbol-scope stores, and the option database). Those stores and the console
//! `IfaceDecompCommand` machinery are a W5+ surface. This module ports the
//! console-independent core that the command is built on:
//!
//!   - [`KunaAssertion`]: the recorded-assertion struct (the session log entry);
//!   - [`AssertLog`]: the session log (the C++ static `assertLog` vector) with
//!     its `kassert list` rendering;
//!   - [`validate_assertion`]: the stage/sub-stage validation + strength
//!     resolution + the **minimal-rewind computation** the command performs
//!     before it dispatches (the half that is independent of the live stores);
//!   - [`Dispatch`]: the sub-stage -> store routing *decision* (which store a
//!     sub-stage routes to, and whether it is latent/unroutable), transcribed
//!     from the C++ `if/else` chain — the actual store mutation is the W5 stub.
//!
//! STUB(W5): the live dispatch (`Override::insertForceGoto`, `FuncProto`
//! locks, `OptionDatabase::set`, symbol retype/rename) and the console parsing
//! of `<args>` land with the console wave on top of [`Dispatch`].

use crate::kuna_phases::{lookup_subphase, KunaPhase, KunaStrength};
use kuna_base::error::{KunaError, KunaResult};

/// (kuna) One recorded typed assertion (session log entry) — C++ `KunaAssertion`.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct KunaAssertion {
    /// Function it applies to, or "(global)".
    pub func_name: String,
    /// Stage asserted.
    pub phase: KunaPhase,
    /// Sub-stage name (validated against the catalog).
    pub subphase: String,
    /// The assertion arguments as given.
    pub args: String,
    /// Requested strength (hard/hint).
    pub strength: KunaStrength,
    /// Strength of the underlying mechanism actually used.
    pub applied: KunaStrength,
    /// Computed minimal rewind stage (REPORTED, not enacted).
    pub rewind: KunaPhase,
}

/// Which existing store a sub-stage's assertion routes to — the *decision*
/// half of the C++ `IfcKunaAssert::execute` dispatch chain.
///
/// The C++ command performs the matching store mutation inline; the Rust port
/// separates the routing decision (here, console-independent) from the store
/// mutation (W5). [`Dispatch::for_substage`] reproduces the C++ `if/else` order
/// exactly, so an unknown/unroutable sub-stage yields the same outcome
/// ([`Dispatch::Latent`] vs [`Dispatch::Unroutable`]) the C++ `else` branch
/// produces.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Dispatch {
    /// -> `Override::insertForceGoto` (S7 edge-virtualization).
    ForceGoto,
    /// -> `Override::insertDeadcodeDelay` (S3 dead-definition-gate).
    DeadcodeDelay,
    /// -> `Override::insertMultistageJump` (S2 switch-model).
    MultistageJump,
    /// -> `Override::insertFlowOverride` (S2 flow-classification).
    FlowOverride,
    /// -> `FuncProto` input/output locks (S4 prototype-source).
    ProtoLock,
    /// -> retype + typelock (S5 type-propagation).
    Retype,
    /// -> `Symbol::setIsolated` (S6 merge-aggressiveness).
    Isolate,
    /// -> rename + namelock (S9 naming-policy).
    Rename,
    /// -> `option <name> ...` (the option database). Carries the option name.
    Option(&'static str),
    /// The sub-stage is LATENT (no override surface exists yet).
    Latent,
    /// The sub-stage is known but not routable through `kassert` (use its
    /// native surface, reported in the sub-stage's `exposure`).
    Unroutable,
}

impl Dispatch {
    /// Route a sub-stage name to its store — verbatim transcription of the C++
    /// `IfcKunaAssert::execute` `if/else` dispatch chain (`kuna_assert.cc:99`),
    /// in the same order. The trailing `else` distinguishes LATENT sub-stages
    /// (`sub->latent`) from merely-unroutable ones, exactly as the C++ does.
    pub fn for_substage(sname: &str, latent: bool) -> Dispatch {
        match sname {
            "edge-virtualization" => Dispatch::ForceGoto,
            "dead-definition-gate" => Dispatch::DeadcodeDelay,
            "switch-model" => Dispatch::MultistageJump,
            "flow-classification" => Dispatch::FlowOverride,
            "prototype-source" => Dispatch::ProtoLock,
            "type-propagation" => Dispatch::Retype,
            "merge-aggressiveness" => Dispatch::Isolate,
            "naming-policy" => Dispatch::Rename,
            "comparison-canonicalization" => Dispatch::Option("compareform"),
            "pointer-notation" => Dispatch::Option("arraynotation"),
            "simplification-quiescence" => Dispatch::Option("togglerule"),
            "readability-rewrites" => Dispatch::Option("analyzeforloops"),
            "extrapop" => Dispatch::Option("extrapop"),
            "pipeline-variant" => Dispatch::Option("setaction"),
            // C++ else branch: LATENT sub-stages report a distinct message from
            // the otherwise-not-routable ones.
            _ => {
                if latent {
                    Dispatch::Latent
                } else {
                    Dispatch::Unroutable
                }
            }
        }
    }
}

/// The result of validating a `kassert <phase> <subphase> [hard|hint]` request,
/// before any store mutation — the console-independent half of
/// `IfcKunaAssert::execute`.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ValidatedAssertion {
    /// The validated stage.
    pub phase: KunaPhase,
    /// The validated sub-stage name.
    pub subphase: String,
    /// The requested strength (the catalog default, or the trailing hard|hint
    /// override).
    pub requested: KunaStrength,
    /// The strength of the underlying mechanism actually used (the sub-stage's
    /// catalog strength — Ghidra stores are uniformly HARD mechanisms).
    pub applied: KunaStrength,
    /// The computed minimal rewind stage (REPORTED; section 12).
    pub rewind: KunaPhase,
    /// Which store the sub-stage routes to.
    pub dispatch: Dispatch,
}

/// Validate a `kassert` request and compute its strength + minimal rewind —
/// the console-independent core of `IfcKunaAssert::execute`
/// (`kuna_assert.cc:51`), up to (but not including) the live store mutation.
///
/// `stagecode` is the `<phase>` token ("P0".."P9", legacy "S1".."S9" accepted); `subname` is the
/// `<substage>` token; `strength_override` is the optional trailing `hard|hint`
/// (the C++ pops it off the token list). Errors mirror the C++ `IfaceParseError`
/// throws (as [`KunaError::Parse`]), in the same precedence order:
///
///   1. bad stage code,
///   2. missing sub-stage name,
///   3. unknown sub-stage,
///   4. sub-stage belongs to a different stage.
///
/// The **minimal-rewind computation** is the sub-stage's catalog `rewind`
/// field (`docs/history/stage-model.md` section 12: format->P9, force-goto->P7, typelock->P5,
/// proto->P4, jumptable/flow/context->P2, deadcode-delay->P3), looked up here.
pub fn validate_assertion(
    stagecode: &str,
    subname: &str,
    strength_override: Option<KunaStrength>,
) -> KunaResult<ValidatedAssertion> {
    let phase = KunaPhase::from_code(stagecode).ok_or_else(|| KunaError::Parse {
        explain: format!("Bad phase code (expecting P0..P9): {stagecode}"),
    })?;
    if subname.is_empty() {
        return Err(KunaError::Parse {
            explain: "Missing sub-phase name (see `phase list`)".to_string(),
        });
    }
    let sub = lookup_subphase(subname).ok_or_else(|| KunaError::Parse {
        explain: format!("Unknown sub-phase: {subname} (see `phase list`)"),
    })?;
    if sub.phase != phase {
        return Err(KunaError::Parse {
            explain: format!(
                "Sub-phase {subname} belongs to phase {}, not {stagecode}",
                sub.phase.code()
            ),
        });
    }

    let requested = strength_override.unwrap_or(sub.strength);

    Ok(ValidatedAssertion {
        phase,
        subphase: subname.to_string(),
        requested,
        applied: sub.strength,
        rewind: sub.rewind,
        dispatch: Dispatch::for_substage(subname, sub.latent),
    })
}

/// The session log of applied assertions — the C++ static `assertLog` vector,
/// reified so it can be owned by the console state (console-lifetime; survives
/// function clears like the Override store does).
#[derive(Clone, Debug, Default)]
pub struct AssertLog {
    entries: Vec<KunaAssertion>,
}

impl AssertLog {
    /// A new, empty log.
    pub fn new() -> AssertLog {
        AssertLog::default()
    }

    /// Record an applied assertion (C++ `assertLog.push_back(rec)`).
    pub fn push(&mut self, rec: KunaAssertion) {
        self.entries.push(rec);
    }

    /// The recorded assertions, in insertion order.
    pub fn entries(&self) -> &[KunaAssertion] {
        &self.entries
    }

    /// true if no assertions are recorded.
    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }

    /// Render `kassert list` — byte-for-byte port of
    /// `IfcKunaAssert::listAssertions` (`kuna_assert.cc:249`). Returns the exact
    /// text the C++ writes to the bulk stream (one trailing newline per line).
    pub fn render_list(&self) -> String {
        let mut os = String::new();
        if self.entries.is_empty() {
            os.push_str("No kuna assertions recorded this session\n");
            return os;
        }
        os.push_str(
            "Recorded assertions (rewind scope is REPORTED per stage-model.md s12; \
             Ghidra-actual re-run remains whole-function):\n",
        );
        for (i, rec) in self.entries.iter().enumerate() {
            os.push_str("  ");
            os.push_str(&i.to_string());
            os.push_str(": [");
            os.push_str(&rec.func_name);
            os.push_str("] ");
            os.push_str(rec.phase.code());
            os.push('/');
            os.push_str(&rec.subphase);
            if !rec.args.is_empty() {
                os.push(' ');
                os.push_str(&rec.args);
            }
            os.push_str("  strength=");
            os.push_str(match rec.strength {
                KunaStrength::Hard => "HARD",
                KunaStrength::Hint => "HINT",
                KunaStrength::None => "-",
            });
            if rec.strength == KunaStrength::Hint && rec.applied == KunaStrength::Hard {
                os.push_str(" (applied via HARD mechanism)");
            }
            os.push_str("  rewind->");
            os.push_str(rec.rewind.code());
            os.push('\n');
        }
        os
    }
}

#[cfg(test)]
mod tests;
